### ISSUE TITLE
HAProxy Cluster Clears Stick Tables Intermittently on Reload

### ISSUE THREAD

**Issue by etfloyd** [haproxy/haproxy#1216]
## Detailed description of the problem
When all members of an haproxy peer cluster are reloaded simultaneously or within a short period of time, occasionally all members will lose stick table entries. ~This does not happen when the cluster consists of a single member.~ The probability of clearing stick tables increases with the size of the configuration file.

## Expected behavior
Stick table contents are reliably preserved across reloads in a multi-member peer cluster. 

## Steps to reproduce the behavior
[sticktabledemo.tar.gz](https://github.com/haproxy/haproxy/files/6268859/sticktabledemo.tar.gz)

1. Un-tar the attached file in an empty directory on a Linux or Mac OS host with Docker installed.
2. Run start.sh and follow the suggestions at the end of its output.

## Do you have any idea what may have caused this?
Since it is intermittent ~and it doesn't happen with a single instance~, I'm guessing it's some kind of race condition between local hand-off during reload and checkpointing between local and remote peer instances. Perhaps a checkpoint is occurring before the new instance created by the reload is quite ready?
Update 4/7/2021: I was able to reproduce the problem with a single instance. It required much more rapid reload cycle and it took a while, but it happened.
## Do you have an idea how to solve the issue?
Not really.

## What is your configuration?
Provided in the attachment.

Here is the top part of an haproxy.cfg file generated by start.sh.
```
global
  stats socket /run/haproxy.sock mode 600 level admin
  log stdout format raw local0
  chroot /root
  user haproxy
  group haproxy

defaults
  log global
  option dontlognull
  mode http
  timeout client 60s
  timeout connect 60s
  timeout server 60s

peers proxies
  bind 127.0.0.1:"${PEERPORT}"
  server haproxy1 "${PEER1}"
  server haproxy2 "${PEER2}"
  table t1 type string len 32 size 512000 expire 25h store server_id

frontend common
  bind 127.0.0.1:"${FRONTPORT}"
  default_backend back1

backend back1
  server back1 127.0.0.1:"${BACKPORT}"

backend back2
... (continues to backend back300)

```
The start.sh script in the demo deploys two instances of haproxy using the same configuration with different environment variables. The frontend and backends are not intended to be used. They are present to add volume to the configuration file for demo purposes. When start.sh is run it outputs the the following at the end:
```
Containers 'haproxy1' and 'haproxy2' are started.
Use 'stop.sh' on the host to stop and remove the containers.
Enter conntainer using, e.g., 'docker exec -it haproxy1 sh'
These scripts only work inside one of the containers:
  /etc/haproxy/show.sh               - shows the number of entries in the table
  /etc/haproxy/reload.sh <sleeptime> - reloads haproxy, sleeps <sleeptime> seconds,
                                       then shows number of entries in the table
Try this:
  while true; do /etc/haproxy/reload.sh 0.10; done
If you use two sessions or tmux with split screen and do this concurrently in
both haproxy1 and haproxy2, the stick table will be empty within a minute or two
```
The configuration file and rapid reload cycle are contrived to make the problem appear relatively quickly. Our actual deployed configuration, of course, is very different from this (see below) and reloaded much less frequently - no more often than 60 seconds - so the problem may not appear for several hours.

Update 4/7/2021: By eliminating the sleep, I was able to reproduce the problem with a single instance.
## Output of `haproxy -vv` and `uname -a`
There is an environment variable in start.sh that allows you to select any of your official (Alpine-based) haproxy images to test with. I have verified that the problem occurs with 2.0.20-alpine, 2.0.21-alpine, 2.0.21 (debian base), 2.2.13-alpine, and 2.3.9-alpine. The demo scripts are written for the alpine versions and may not work correctly on the debian base, but we have tested with non-alpine images and there's no difference in behavior.

```
(paste your output here)
```

## If HAProxy crashed: Last outputs and backtraces
No crash.

```
(paste your output here)
```

## Additional information (if helpful)
So, a brief description of the use case that brought us here. We have implemented an haproxy-based dynamically configured load balancer layer that feeds upstream requests to multiple kubernetes clusters. This is in an environment that does not support native LoadBalancer-type Kubernetes services. So, the automation monitors the Kubernetes clusters and uses the data plane API to manage the haproxy configurations to provide a similar capability. The layer is implemented as a peer cluster with three instances for high availability and it worked just fine until we onboarded an application with a requirement for cookie-based session persistence, and during the same time period, we had some node instability on one of the clusters. This caused the configuration to change frequently and data plane api to reload the haproxys frequently and more or less simultaneously and this is what surfaced the problem. The configuration files that the automation builds are fairly large: ~2,000+ lines, 300 back ends. Not huge, but still larger than the manually-maintained configurations we've previously deployed. We found that the larger the configuration file, and the more frequently it is reloaded, the more likely we are to lose stick tables. This is reflected in the 'start.sh' demo script in the attached archive.

**Comment by EmericBr** [comment]
Hi Etfloyd, are you sure to not reload too frequently. During a reload the new process will request the old one for a full resync, but it you are performing an other reload before the end of this full initial resync, it will result to the lost of the entry not yet learned by the second process at that point. It's a known limitation of the current protocol implementation. Current peer implementation is not compatible with hard frequency reloading.

**Comment by etfloyd** [comment]
Interesting. First I've heard about this limitation. I must have overlooked it somehow in the documentation. ;)

Very high-frequency reloading was done for the demo just for the purpose of recreating the issue without having to wait for hours. In the actual deployment, reload frequency is determined by the rate of change in the downstream clusters and triggered by the Data Plane API on 'commit' whenever it decides a reload is necessary. We don't actually know in advance whether it will do that or not, but we scrape the clusters at 60-second intervals so I know it's no more often than that. I didn't really consider 60-second intervals to be "high-frequency".

Is there a reliable way to detect when a reload is in progress? I may be able to code around this limitation by delaying calls to the DP API until a previous reload is complete.

**Comment by EmericBr** [comment]
After checking if a new process fails to retrieve a full resync from an old one,.  It will request a full resync to a remote node. But it could also fail if remote node have been also reloaded, or if you have only one node. This way to process is very old, older than the "master" mode of haproxy, I don't know the interaction with that and I should ask to @wlallemand

**Comment by EmericBr** [comment]
"III) Initial full resync process chapter" in doc/peers-v2.0.txt describes this process, this documentation may not be fully updated but this process is inherited from peers v1.0 (commit 2b920a1af). This process worked at this time because the users were reloading manually haproxy but it is clearly not designed for nowadays usage and deserve to be re-worked.

**Comment by EmericBr** [comment]
I've several pending fixes on resync process, I hope it will fix your issue but what you observe hardly reloading haproxy may not be the same issue than the one appearing sometimes

**Comment by etfloyd** [comment]
Ah, that doc explains a lot! I was looking in the public docs [here](http://www.haproxy.org/#docs), not the internals docs in the source repo.

For our specific application, I believe I can work around this limitation for now simply by serializing reloads. For that, I need a 100% reliable way to tell if a reload is in progress. The automation runs inside the container with the haproxy instance and the DP API, so I can query the stats socket or the master cli socket, or both if needed. What do you suggest?

**Comment by EmericBr** [comment]
We should ask @wlallemand but I think we could have the info checking if there is remaining old processes listing processes on master socket.

**Comment by EmericBr** [comment]
But an old process could remain up until it finish to serve all its sessions, which could take a long time. I will check if we can have the info of full resync finished showing the "show peers" on cli.

**Comment by EmericBr** [comment]
You could already check if full resync process is ended and know that you are free to perfom a reload checking some flags on the return of the cli command "show peers" but it shows internal flags which could evolve in further version. WIlly and I talk today to add a more "generic" way to show this information which survives versions upgrade. 

But for now if you want to test "show peers"command cli will show those lines for each peer sections:
0x5590cd54e270: [08/Apr/2021:14:30:37] id=sharedlb state=0 flags=0x3 resync_timeout=<PAST> task_calls=12

and here if the flags 0x1 and 0x2 are present ( this means ( (flags & 0x3) == 0x3 ), it means the current process has tried every thing to get a initial full resync and is now conisdering runtime updates (so you could consider your are free to perform a reload at this step).

**Comment by etfloyd** [comment]
I'm trying it out with this modified reload.sh script in the demo setup:
```
/etc/haproxy # cat reload.sh 
#!/bin/sh
# Reloads haproxy, sleeps for $1 seconds (default: 1s), displays the table element counts
# This script works inside an alpine-based haproxy container
NAP="${1:-1}"
if echo "show peers" | nc local:/run/haproxy.sock | grep 'proxies' | grep 'flags=0x3' >/dev/null; then
  if [ -f /tmp/notready ]; then
    sleep 1
    rm -f /tmp/notready
  fi
  kill -USR2 1
else
  echo "show peers" | nc local:/run/haproxy.sock | grep 'proxies' | grep 'flags='
  touch /tmp/notready
fi
sleep $NAP
echo "show table" | nc local:/run/haproxy.sock
```
Running this inside both containers:
```
while true; do /etc/haproxy/reload.sh 0.01; done
```
Before it would have completely lost stick table contents within few seconds. Now it's been running for a while and cookies are still there. I can see the 5-second timeouts in action occasionally - one or the other will show a count down on resync_timeout.

I have to leave the house for a couple of hours. I'll leave it running and see if cookies are still there when I return. In any case, it's a vast improvement. Thank you!

**Comment by etfloyd** [comment]
Well, It didn't survive the two hours I was gone. I noticed something: One of the containers was "stuck" with flags=0x0 which means that it must have reloaded once and never actually completed a full resync. Here's the output from one of the thousands:
```
0x7f2cb1e07e60: [08/Apr/2021:13:51:59] id=proxies disabled=0 flags=0x0 resync_timeout=<NEVER> task_calls=26368
# table: proxies/t1, type: string, size:512000, used:0
```
There are thousands of line pairs like this where task_calls increments every few but nothing else changes. The timestamp indicates that it went into this state around 20 minutes after I started the 'while' so it definitely survived a lot longer than before. The output from the other container looked normal, except it also showed 0 stick table entries.

**Comment by etfloyd** [comment]
The modified reload.sh script prevents a reload until 1 second after recovering from any occurrence of flags != 0x3 which means that once it got stuck no further reloads were triggered. On the one that was stuck, I stopped the script and triggered a reload manually with kill -USR2 1 and it recovered. Of course, the stick table entries were all gone by that time but it at least it was no longer stuck.

**Comment by etfloyd** [comment]
Ah, I'm rerunning the longer time test because I had previously made a change to the configuration and inadvertently left it in. I had added:
```
mworker-max-reloads 2
```
to the global section while trying various ways on my own to determine if a reload was in progress. I've taken that line out now and so far it's working fine.

**Comment by etfloyd** [comment]
This time it ran 33 minutes before trashing the stick tables. I put timestamps on all the output from the script so I could locate the exact point of failure and I think I see what happened. It looks like both containers triggered a reload at almost exactly the same time. Then I see both count down resync_timeout from 4 or 5s to 0s twice, the first time with flags=0x0, the second with flags=0x1. From peers-v2.0.text, these count downs must be the 5 second timeouts mentioned in III) a). Looks like neither container received a connection from its local process and neither one found the other to be available for a resync, so both gave up and left the tables empty. Notably, neither one got "stuck" this time.

Looking further through the logs I see a number of instances where one or the other went through a resync_timeout count-down one time with flags=0x0 but recovered after that. I'm guessing that it failed to get a connection from its local process but was able to resync from the other one. (This begs the question, what happened to the local process? Why would it have ended before syncing stick tables?)

Given the extreme high frequency of reloads, 33 minutes is actually pretty good. I need to test some more with larger config files to get some idea of what the probability of failure in the actual use case might be. Those failures are visible to the customer and attract quite a bit of management attention, so we still need a permanent fix, but I may be able to construct a "good enough" temporary work-around with this.

**Comment by etfloyd** [comment]
Update - I have a workaround that seems to work. I've pasted in the script in case it might be useful to others. There are four cases that have to be handled:

(Retrieve the sync state using "show peers".)
1. Stuck: flags != 0x3 and resync_timeout=`<NEVER>` - The only way out of this seems to be to trigger a reload, so do that now.
2. Broken: flags == 0x1 and resync_timeout is counting down - Resync from local has failed and it's now trying to resync from a remote peer and failing. We're likely to lose data. Set the 'broken' flag.
3. NotReady: flags == 0x0 and resync_timout is counting down - It's trying to resync from local and failing. This is OK as long as it can find a healthy remote peer to resync from.
4. Ready: flags == 0x3 - Resync is complete. It should be OK to trigger a reload. If the 'broken' flag is set, restore stick table entries from disk otherwise save them to disk. Trigger a reload.

I set up a three-container haproxy cluster, expanded the config file to ~2,400 lines and 800 backends, then ran the revised script in a loop on each doing 5-10 reloads / second for a minimum of 60,000 reloads on each container and it never lost stick table data.

```
#!/bin/sh
# Reloads haproxy preserving stick table contents
# This script works inside an alpine-based haproxy container

STATS_SOCKET="/run/haproxy.sock"
SAVE_PATH="/etc/haproxy/stickdata/$HOSTNAME"
SAVE_WORK="tempfile_delete_me"
PEERS_SECTION="proxies"

debug() {
  >&2 echo "$(date '+%F %T') DEBUG: $1"
}

# Sends $1 command to haproxy stats socket
haproxy_command() {
  echo "$1" | nc "local:$STATS_SOCKET"
}

# Display table counts or contents of table $1
show_table() {
  haproxy_command "show table $1"
}

# Display peers info
show_peers() {
  haproxy_command "show peers" | grep "$PEERS_SECTION" | grep 'flags=0x'
}

# Saves the $1 table to disk in a format that can be reloaded directly into the stats port
save_table() {
  local COUNT
  local TABLE="$1"
  local FILEPATH="$SAVE_PATH/$(echo $TABLE | tr '/' '_')"
  local WORK="$SAVE_PATH/$SAVE_WORK"
  [ -z "$TABLE" ] && return 0
  mkdir -p "$SAVE_PATH"
  show_table "$TABLE" \
    | awk '/key=/ {print "set table '"$TABLE"'", $2, "data."$5 }' \
    | tr '=' ' ' \
    >$WORK
  COUNT="$(cat $WORK | wc -l)"
  [ ! "$COUNT" = "0" ] && mv -f $WORK $FILEPATH || rm -f $WORK
}

# Saves all tables with non-zero entries to disk
save_tables() {
  local line
  show_table | while read line; do
    local COUNT="$(echo "$line" | awk '{ print $7 }' | cut -d ',' -f1 | cut -d ':' -f2)"
    local TABLE="$(echo "$line" | awk '{ print $3 }' | cut -d ',' -f1 )"
    [ -z "$TABLE" ] && continue
    [ "$COUNT" = "0" ] && debug "Table $TABLE Count $COUNT" || save_table "$TABLE"
  done
}

# Restores $1 table previously saved by save_table()
restore_table() {
  local TABLE="$1"
  local FILEPATH="$SAVE_PATH/$(echo $TABLE | tr '/' '_')"
  if [ -f "$FILEPATH" ]; then
    cat "$FILEPATH" | nc local:$STATS_SOCKET >/dev/null
    local COUNT="$(cat $FILEPATH | wc -l)"
    debug "$COUNT entries for table $TABLE restored from $FILEPATH"
  fi
}

# Restores all tables previously saved by save_tables()
restore_tables() {
  local line
  show_table | while read line; do
    local TABLE="$(echo "$line" | awk '{ print $3 }' | cut -d ',' -f1 )"
    [ -z "$TABLE" ] && continue
    restore_table "$TABLE"
  done
}

# Returns true if reload initial resync is NOT active
check() {
  show_peers | grep 'flags=0x3' >/dev/null && return 0
  # try one more time after a very brief sleep ( prevents many spurious not-readys )
  sleep "0.0$RANDOM"
  show_peers | grep 'flags=0x3' >/dev/null && return 0
  return 1
}

# Returns true if reload inital resync is probably stuck
stuck() {
  show_peers | grep -v 'flags=0x3' | grep 'NEVER' && return 0
  return 1
}

# Returns true if initial sync is probably going to lose stick table data
broken() {
  show_peers | grep 'flags=0x1' && return 0
  return 1
}

# Triggers a reload
trigger_reload() {
  kill -USR2 1 # haproxy master pid
}

# Return true if resync is NOT complete
wait_ready() {
  check && return 1
  # Are we broken or stuck?
  broken && touch /tmp/broken
  if stuck; then
    # Initial resync is probably stuck - knock it loose with another reload
    debug "Recovering from stuck resync!"
    trigger_reload
    sleep 0.1
  else
    # Not stuck, so we're just waiting on resync to finish
    sleep 0.5
  fi
  return 0
}

# Fix broken tables if necessary. Return true if tables needed fixing
fix_broken() {
  if [ -f /tmp/broken ]; then
    # Table data is probably lost
    debug "Recovering from broken resync!!"
    restore_tables
    rm -f /tmp/broken
    return 0
  fi
  return 1
}

[ -f /tmp/count ] && COUNT="$(cat /tmp/count)" || COUNT=0 # added for demo testing only

# Skip the workaround if there are no tables or no peers
if [ -z "$(show_table)" ] || [ -z "$(show_peers)" ]; then
  trigger_reload
  exit 0
fi

while wait_ready; do continue; done
fix_broken || save_tables
trigger_reload
COUNT="$(( $COUNT + 1 ))"     # added for demo testing only
while wait_ready; do continue; done
fix_broken

debug "$COUNT: $(show_table)" # added for demo testing only
echo "$COUNT" >/tmp/count     # added for demo testing only
```
2021/04/12: Updated script to be more like what we actually deployed

**Comment by EmericBr** [comment]
Hi, there is 2 bunches of fixes recently commited into master concerning the peers and sitck tables.

**Comment by etfloyd** [comment]
HAProxy 2.2.14 has ten peers-related patches and they don't completely fix this issue but they definitely improve it to the point that we may be able to do without my work-around. I tested with the original sticktabledemo scripts attached above, slightly modified to display an iteration counter, with start.sh updated to specify HAProxy version 2.2.14, and with 800 backends, running this inside each container: 
```
while true; do /etc/haproxy/reload.sh 0.10; done
```
The '0.10' specifies a time interval to sleep between reloads. (Yes, this is _very_ high-frequency - ~10 reloads / sec!) This failed (lost all stick table entries) after a few hundred iterations. Running with sleep time 0.20 (~5 reloads / sec) also failed after 7,000 iterations. Sleep time 0.30 (~3 reloads / sec) ran 40,000 iterations with no data loss before I stopped it and repeated the test with HAProxy version 2.2.13. On 2.2.13 at 3 reloads / sec all stick table entries were gone from both instances after 2,400 iterations.

So, I think we can just put a one-second delay after each reload just to be sure they don't occur in too rapid succession and be pretty comfortable that stick table data will be preserved. In the coming week or so, we'll be testing that in our actual use case.

**Comment by EmericBr** [comment]
Yep, as previously said, the initial resync of stick table is not designed to support such intensive reload, so processing 40 000 iteration at 3 reload/sec without data loss seems reasonable in my opinion to consider the issue fixed and working as designed. So @wtarreau we may close that bug.

**Comment by wtarreau** [comment]
OK. I'll mark it fixed since your patches were backported to 2.2 already. Thanks!
